extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java

```diff
@@ -0,0 +1,220 @@
+
+import javax.annotation.Nullable;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+@JsonTypeName("namespace")
+public class NamespaceLookupExtractorFactory implements LookupExtractorFactory
+{
+  private static final Logger LOG = new Logger(NamespaceLookupExtractorFactory.class);
+
+  private static long SCHEDULE_TIMEOUT = 60_000;
+
+  private final AtomicBoolean started = new AtomicBoolean(false);
+  private final ReadWriteLock startStopSync = new ReentrantReadWriteLock();
+  private final ExtractionNamespace extractionNamespace;
+  private final NamespaceExtractionCacheManager manager;
+  private final LookupIntrospectHandler lookupIntrospectHandler;
+
+  private final String extractorID;
+
+  @JsonCreator
+  public NamespaceLookupExtractorFactory(
+      @JsonProperty("extractionNamespace") ExtractionNamespace extractionNamespace,
+      @JacksonInject NamespaceExtractionCacheManager manager
+  )
+  {
+    this.extractionNamespace = Preconditions.checkNotNull(
+        extractionNamespace,
+        "extractionNamespace should be specified"
+    );
+    this.manager = manager;
+    this.extractorID = buildID();
+    this.lookupIntrospectHandler = new LookupIntrospectHandler() {
+      @GET
+      @Path("/keys")
+      @Produces(MediaType.APPLICATION_JSON)
+      public Response getKeys()
+      {
+        return Response.ok(getLatest().keySet().toString()).build();
+      }
+
+      @GET
+      @Path("/values")
+      @Produces(MediaType.APPLICATION_JSON)
+      public Response getValues()
+      {
+        return Response.ok(getLatest().values().toString()).build();
+      }
+
+      @GET
+      @Produces(MediaType.APPLICATION_JSON)
+      public Response getMap()
+      {
+        return Response.ok(getLatest()).build();
+      }
+
+      private Map<String, String> getLatest()
+      {
+        return ((MapLookupExtractor)get()).getMap();
+      }
+    };
+  }
+
+  @Override
+  public boolean start()
+  {
+    final Lock writeLock = startStopSync.writeLock();
+    writeLock.lock();
+    try {
+      if (!started.compareAndSet(false, true)) {
+        LOG.warn("Already started!");
+        return false;
+      }
+      if (!manager.scheduleAndWait(extractorID, extractionNamespace, SCHEDULE_TIMEOUT)) {
+        LOG.warn("Failed to schedule lookup [%s]", extractorID);
+        return false;
+      }
+      LOG.debug("NamespaceLookupExtractorFactory[%s] started", extractorID);
+      return true;
+    }
+    finally {
+      writeLock.unlock();
+    }
+  }
+
+  @Override
+  public boolean close()
+  {
+    final Lock writeLock = startStopSync.writeLock();
```

Comment:
I'd rather not if possible. there's no reason to force the get() to be synchronized.